### ISSUE TITLE
Ubuntu 22.04 LTS guest, possible incorrect permissions for shared folders

### ISSUE THREAD

**Issue by tomearp** [vmware/open-vm-tools#602]
### Describe the bug

TL;DR I don't appear to have write access to the root of a shared folder, only to files and subfolders in the shared folder.
I don't know if this is expected behaviour?

VMware Worksation 16 Pro, Win10 host, Ubuntu 22.04 LTS guest.
Installed versions of open-vm-tools/-desktop:
```
open-vm-tools-desktop/jammy,now 2:11.3.5-1ubuntu4 amd64 [installed]
open-vm-tools/jammy,now 2:11.3.5-1ubuntu4 amd64 [installed]
```
Shared folders automounted by fstab entry:
```
`.host:/ /mnt/hgfs fuse.vmhgfs-fuse auto,allow_other 0 0`
```
If I open a shared folder in Files, then in the root of the shared folder I can't carry out operations such as Delete, New Folder, Cut, Move to, Rename. All operations work within subfolders.

Looking at the permissions in the terminal, write access is not given to the shared folder (Z in this case), but files/folders within the shared folder are 777:
```
moo@moo-virtual-machine:/mnt/hgfs$ ls -al
total 29
dr-xr-xr-x 1 root root  4192 Jul 13  2022 .
drwxr-xr-x 3 root root  4096 Jul 13 10:53 ..
dr-xr-xr-x 1 root root 20480 Jul 13 11:26 Z
```
```
moo@moo-virtual-machine:/mnt/hgfs/Z$ ls -al
ls: cannot access 'System Volume Information': Permission denied
total 1148
dr-xr-xr-x 1 root root      0 Jun  4 17:17 '$RECYCLE.BIN'
drwxrwxrwx 1 root root      0 Jul 13 11:38  apps
drwxrwxrwx 1 root root 303104 Jun  7 13:40  iplayer
drwxrwxrwx 1 root root   4096 Jul 13 11:37  media
drwxrwxrwx 1 root root      0 Jun 23 14:47  sab
drwxrwxrwx 1 root root   4096 Jul 13 11:38  switch
d????????? ? ?    ?         ?            ? 'System Volume Information'
-rwxrwxrwx 1 root root 863437 Jun 24 13:16  unencrypted.pdf
```

### Reproduction steps

```bash
1. Make a Ubuntu 22.04 LTS VM in VMware Workstation 16 Pro.
2. Mount your shared folders.
3. Using Files, try and create a new folder, or delete a file, in the root of a shared folder.
```

### Expected behavior

Full permissions in the shared folder.

**Comment by PaTHml** [comment]
Hi Tomearp,

Thank you for reporting the issue. An internal issue was created.
In the mean time was there any error reported when enabling the shared drive?

**Comment by PaTHml** [comment]
I've installed VMware Workstation 16.2.3 Pro and a base Ubuntu 22.04 LTS (from iso). My host however is Windows 11, and I did not see the issue.

Can you share more version details:
- VMware Workstation's complete version (from help->About VMware Workstation)?
- Windows 10 version (from Settings->System->About)? (is is 21H2 etc.)

The way I created the share was like this (maybe I'm wrong):
- Created a C:\vm-shares\A path in host,
- Added path in the VM settings->options->Shared Folders  and set to "Always enabled"
- powered on the VM.
- Created a few files and subfolder in the A path from the windows side.
- From terminal: "sudo mkdir -p /mnt/hgfs/A"
- From terminal: "sudo vi /etc/fstab"
- Added a line: ".host:/A /mnt/hgfs/A fuse.vmhgfs-fuse allow_other,uid=<output-of-id-u>,gid=<output-of-id-g> 0 0"
- reboot

I see 777 for the A folder whit ls -l /mnt/hgfs; the owner being my user (with provided uid and gid values in fstab).

Is that close to what you did to produce the issue?

**Comment by pengzhencao** [comment]
From the Folder list in folder "Z" I could see there is Recycle.BIN and "System Volume Information", I assume you are sharing the entire disk of your windows host to the VM. I guess this might be the cause for your problem. Could you try to share a normal host directory rather than the entire driver and see if it could workaround your issue?

**Comment by PaTHml** [comment]
@tomearp, if 'Z' is mapped to hosts "somedrive-letter:\"; the host access rules are driving the top level permissions.
There's a few questions to answer in the previous posts.

Thanks,

**Comment by tomearp** [comment]
@pengzhencao yes, I am sharing an entire drive but is not a system drive; I could unplug it and it would have no effect on the host's operation.

Is this a protection mechanism within open-vm-tools, or VMware Workstation, to prevent people sharing their system drive with a VM that could potentially do bad things?

It would appear to be something like that, because if I share a subfolder of the drive in addition to the root, then the subfolder share is given 777 and Files works normally:
```
moo@moo-virtual-machine:/mnt/hgfs$ ls -al
total 29
dr-xr-xr-x 1 moo  moo   4192 Jul 15 21:28 .
drwxr-xr-x 3 root root  4096 Jul 13 10:53 ..
drwxrwxrwx 1 moo  moo      0 Jun 23 14:47 sab
dr-xr-xr-x 1 moo  moo  20480 Jul 13 21:45 Z
```

I understand why such a thing would limit access to the system drive of the host. I don't know if there is any mechanism within VMware/VirtualBox etc. to be able to tell open-vm-tools which host drives are system drives and which are just 'general' storage?

At least I can work around it by moving everything into a subfolder and mapping that instead.

**Comment by PaTHml** [comment]
Looking at the fstab entry:
".host:/  /mnt/hgfs ..."

2 things jump out (I'll need to try them out too).
1. The Worstation->VM->Settings->Options->Shared folders->Folders has "name" and "shared path" fields. Usually the "name" value is used after the ".host:" written as a path, "/A"  in my example. does that mean you have an empty name field
2. the mount point is /mnt/hgfs directly.  Rather than a subfolder in /mnt/hgfs.


I don't think the "auto" option on the mount should impact anything but that one last permutation to explore.

Thank you for the details!

**Comment by tomearp** [comment]
Not specifying a name after the .host: part means it will mount all enabled shared folders.

If you want to mount each shared folder differently, then you can create separate fstab entries for each of them. But I just mount them all to the same point, so I just have the single entry that mounts everything at once.
<br>
I don't think there's anything special about /mnt/hgfs, it's just a folder I created as a mount point. if I change the mount point the behaviour remains the same.
<br>
I also tried changing the mount options in the fstab entry. I tried `defaults,allow_other` and `allow_other` on its own. The permissions remain the same for the shared folder that is linked to an entire drive.
<br>
I also experimented with creating an additional mount point on the host for the Z drive. As well as the Z letter assignment, I mounted it to a folder called test on the X drive, and created a shared folder entry for it in the VM settings. This works fine, and is a better workaround for me, as I don't need to disrupt the existing folder structure on the Z drive. I can just rename the shared folder entry to Z and anything in the guest that is expecting to see /mnt/hgfs/Z with 777 permissions on everything will get it.
```
moo@moo-virtual-machine:/mnt/hgfs$ ls -al
total 189
dr-xr-xr-x 1 root root   4192 Jul 16  2022  .
drwxr-xr-x 3 root root   4096 Jul 13 10:53  ..
drwxrwxrwx 1 root root 163840 Nov 10  2020 'America 2017'
drwxrwxrwx 1 root root      0 Jul 16 13:05  test
dr-xr-xr-x 1 root root  20480 Jul 15 21:56  Z
```

![2022-07-16T120804Z](https://user-images.githubusercontent.com/17062157/179354412-43734797-8aca-42eb-a3dd-27eb5a94e323.png)

![2022-07-16T120829Z](https://user-images.githubusercontent.com/17062157/179354415-2e740e11-2d75-478a-bb9b-a99979e0f0ec.png)

**Comment by tomearp** [comment]
Just a quick post to confirm exactly what is happening and the workarounds.

Trying to share an entire drive from the host results in the root of the shared folder not being writeable by Files in ubuntu (meaning you can't move, delete, cut, make a new folder within the Files app). Subfolders of the shared folder work as expected (777 permissions).
Other methods of writing to files in the root of the folder do appear to work (e.g. nano, Text Editor). Creating folders with mkdir works.

The workarounds are:
1. Don't share an entire host drive, share a folder from the drive.
2. Create an additional mount point on the host that points to the drive (i.e. map it to an NTFS folder on another drive within Computer Management). Create a shared folder in the guest settings that points to this mount point.
![2022-07-16T122002Z](https://user-images.githubusercontent.com/17062157/179354877-834bbf12-796e-4053-9dd1-c17327c4e1dd.png)
![2022-07-16T122018Z](https://user-images.githubusercontent.com/17062157/179354879-6e30e019-fe83-4f2f-bf0d-7046a97a9b56.png)

**Comment by PaTHml** [comment]
From what I tried (using a USB thumb drive (as 'D:')),
Setting up the Folder share as (name=/, path=D:\) and using ".host:/" in fstab had a folder showing up as /mnt/hgfs/'!%' .
Setting up the folder share as (name=D, path=D:\) and using ".host:/" in fstab had a folder showing up as /mnt/hgfs/D . 

In my case (not the same) permissions were usable, allowing me to edit files in the thumb drive's root.

Seems USB drive is not a good repro solution; I need something internal (sata) and to enable bitlocker as well.
Let you know what I find. 

Unless I missed it, you have not provided the full VMware Worstation version yet. If you have time; might be something different version to version. Same for Windows 10.

**Comment by tomearp** [comment]
sorry, version details of everything below.

Host OS: Microsoft Windows 10 Version 21H2 (OS Build 19044.1826)

Guest OS: Ubuntu 22.04 LTS, kernel 5.15.0-41-generic

VMware Workstation 16 Pro, version 16.2.3 build-19376536

open-vm-tools-desktop/jammy,now 2:11.3.5-1ubuntu4 amd64 [installed]
open-vm-tools/jammy,now 2:11.3.5-1ubuntu4 amd64 [installed]
